### PR TITLE
udevadm: add new "udevadm info --tree" command

### PR DESCRIPTION
**Pull request by poettering**
This shows the device hierarchy as a tree, i.e. as what it actually is.

### CODE REVIEW

File: src/udev/udevadm-info.c [hunk 1]
```diff
@@ -486,6 +497,201 @@ static int help(void) {
...
+                r = sd_device_enumerator_add_match_parent(e, below);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to install parent enumerator match: %m");
+        }
+
+        r = sd_device_enumerator_allow_uninitialized(e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable enumeration of uninitialized devices: %m");
+
+        FOREACH_DEVICE(e, child) {
```
Comment (yuwata) [inline src/udev/udevadm-info.c]: This may not show e.g. driver device??

**Review by yuwata — COMMENTED** [review]
The logic of the last commit looks complicated for me. And IIUC the comment in the code, I think it is better to fix the enumerator to actually enumerate all devices??
The enumerator now sorts the devices which mostly fits for showing in tree format.

**Comment by poettering** [comment]
As discussed elsewhere this new rework includes any sysfs kobjects that aren't linked from /sys/class or /sys/bus also in the enumerations if they are parent devices of devices found that way.

Various other fixes.

File: src/udev/udevadm-info.c [hunk 2]
```diff
@@ -486,6 +497,156 @@ static int help(void) {
...
+                if (r < 0)
+                        return r;
+
+                i = j;
+        }
+
+        return 0;
+}
+
+static int print_tree(sd_device* below, const char *prefix, unsigned level) {
```
Comment (yuwata) [inline src/udev/udevadm-info.c]: The second and third arguments are never used.

**Comment by poettering** [comment]
fixed the issue @yuwata pointed out.

**Comment by yuwata** [comment]
Hm, all full-task CIs are red...

**Comment by poettering** [comment]
ah, the pitfalls of log_trace()...

**Comment by poettering** [comment]
Ci should now be fixed. ptal

**Review by yuwata — APPROVED** [review]
LGTM.

**Pull request merged**